python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,539 @@
+
+                    # Fail this concurrent group of testers
+                    for this_job in concurrent_jobs:
+                        tester = this_job.getTester()
+                        tester.setStatus('OUTFILE RACE CONDITION', tester.bucket_fail)
+                        failed_or_skipped_testers.add(tester)
+
+                    # Break out of initial concurrent_jobs and allow checking of the next concurrent group
+                    break
+
+                output_files_in_dir.update(output_files)
+
+            # Delete this group of job containers and allow the loop to continue
+            for job_container in concurrent_jobs:
+                temp_dag.delete_node(job_container)
+
+        # Return any failed or skipped testers discovered
+        return failed_or_skipped_testers
+
+    # Loop through the testers and generate a Tester Data container (which we will call a 'job') which we will pass to the thread
+    # pools for processing
+    #
+    # TODO: I tried to keep the names 'tester' and 'job' separate when working with the tester container (job) or the actual
+    #       tester object (tester) Do better? Its the 'job' object that gets passed between the thread pools.
+    def schedule(self, testers):
+        # Increment our simple queue count with the number of testers the scheduler received
+        with self.thread_lock:
+            self.job_queue_count += len(testers)
+
+        # Create the job DAG object we will share between this group of testers
+        job_dag = dag.DAG()
+
+        # Create a local storage of tester names to job container map
+        name_to_job_container = {}
+        for tester in testers:
+            name_to_job_container[tester.getTestName()] = TesterData(tester, job_dag, self.options)
+
+        # Discover failures only prevalent when scrutinizing the entire group as a whole
+        # Note: this also returns non-failing skipped testers, because its necessary to know about them when discovering things like
+        #       race conditions. It may not belong in that method, but its efficient to perform "while we're here".
+        skipped_or_failed_testers = self.checkGroupFailures(name_to_job_container)
+
+        # Local storage of skipped/failed tester containers (AKA job containers)
+        non_runnable_jobs = set([])
+
+        # Create a set of failing tester data containers
+        for failed_tester in skipped_or_failed_testers:
+            non_runnable_jobs.add(name_to_job_container[failed_tester.getTestName()])
+
+        # Create edge nodes (dependencies) in the DAG shared between this group of testers
+        self.createEdgeNodes(name_to_job_container)
+
+        # Iterate over the items in our non_runnable_jobs and handle any skipped dependencies
+        # Note the copy method, as this may add more skipped tests to the original set as we learn about failed downstream jobs
+        for job in non_runnable_jobs.copy():
```

Comment:
Clever, maybe too clever... I've been down this road before and know why this is here.